src/api-umbrella/utils/interval_lock.lua

```diff
@@ -0,0 +1,78 @@
+local lock = require "resty.lock"
+
+local _M = {}
+
+--- Only one thread can execute fn through its execution duration
+-- A lock (mutex) is held while the function executes and released at the end
+-- @param name - a unique identifier for this lock (automatically namespaced)
+-- @param fn - function to execute if the lock can be held
+-- @return ok, error
+_M.mutex_exec = function(name, fn)
+  local check_lock = lock:new("locks", {["timeout"] = 0})
+  local _, lock_err = check_lock:lock("mutex_exec:" .. name)
+  if lock_err then
+    return true, nil
+  else
+    local results = pcall(fn)
+    local ok, unlock_err = check_lock:unlock()
+    if not ok then
+      ngx.log(ngx.ERR, "failed to unlock: ", unlock_err)
+    end
+    return results
+  end
+end
+
+--- Only allow a function to be executed once in a given interval
+-- A lock (mutex) is set to expire after an interval of time. If the mutex is
+-- present, execution won't take place
+-- @param name - a unique identifier for this lock (automatically namespaced)
+-- @param interval - the length of time until expiry (seconds)
+-- @param fn - function to execute within the interval
+-- @return ok, error
+_M.timeout_exec = function(name, interval, fn)
+  local ok, err = ngx.shared.interval_locks:add(name, true, interval)
+  if not ok and err == "exists" then
+    return true, nil
+  elseif not ok then
+    return false, err
+  else
+    return pcall(fn)
```

Comment:
Same as https://github.com/NREL/api-umbrella/pull/191/files#r45168023 I might be inclined to check the status of this `pcall` and log failures. However, this one is probably the more important one to log, since this is the one that's actually going to be calling our target function in the app, where I'm assuming it's more likely things could unexpectedly break. The other pcalls in this library would just break if something in this library unexpectedly broke, so it seems less likely, but logging failures for all of these still might not be a bad idea.